src/event.rs

```diff
@@ -1,59 +1,411 @@
+            in_callback: self.in_callback.clone(),
+            stopped: self.stopped.clone(),
+        }
+    }
+}
+
+impl<T> From<EventInner<T>> for Event<Internal<T>> {
+    fn from(inner: EventInner<T>) -> Self {
+        Event {
+            inner: Internal(inner),
+            in_callback: Arc::new(AtomicBool::new(false)),
+            stopped: Arc::new(AtomicBool::new(false)),
+        }
+    }
+}
+
+impl<T> From<EventInner<T>> for Event<Local<T>> {
+    fn from(inner: EventInner<T>) -> Self {
+        Event {
+            inner: Local(Rc::new(RefCell::new(inner))),
+            in_callback: Arc::new(AtomicBool::new(false)),
+            stopped: Arc::new(AtomicBool::new(false)),
+        }
+    }
+}
+
+impl<T> From<EventInner<T>> for Event<Sync<T>> {
+    fn from(inner: EventInner<T>) -> Self {
+        Event {
+            inner: Sync(Arc::new(Mutex::new(inner))),
+            in_callback: Arc::new(AtomicBool::new(false)),
+            stopped: Arc::new(AtomicBool::new(false)),
+        }
+    }
+}
+
+impl<T> Event<Internal<T>> {
+    fn with_inner<F, O>(&mut self, f: F) -> O
+    where
+        F: Fn(&mut EventInner<T>) -> O,
+    {
+        let ev = &mut self.inner.0;
+        f(ev)
+    }
+
+    pub fn stop(&mut self) -> io::Result<()> {
+        self.set_stopped(true);
+        self.with_inner(|inner| inner.stop())
+    }
+}
+
+impl<T> Event<Local<T>> {
+    fn with_inner<F, O>(&self, f: F) -> O
+    where
+        F: Fn(&mut EventInner<T>) -> O,
+    {
+        let mut ev = self.inner.0.borrow_mut();
+        f(&mut *ev)
+    }
+
+    pub fn stop(&mut self) -> io::Result<()> {
+        self.set_stopped(true);
+        self.with_inner(|inner| inner.stop())
+    }
+}
+
+impl<T> Event<LocalWeak<T>> {
+    fn with_inner<F, O>(&self, f: F) -> O
+    where
+        F: Fn(&mut EventInner<T>) -> O,
+    {
+        let upgraded = self.inner.0.upgrade().unwrap();
+        let mut ev = upgraded.borrow_mut();
+        f(&mut *ev)
+    }
+
+    pub fn stop(&mut self) -> io::Result<()> {
+        self.set_stopped(true);
+        self.with_inner(|inner| inner.stop())
+    }
+}
+
+impl<T> Event<Sync<T>> {
+    fn with_inner<F, O>(&self, f: F) -> O
+    where
+        F: Fn(&mut EventInner<T>) -> O,
+    {
+        let mut ev = self.inner.0.lock().unwrap();
+        f(&mut *ev)
+    }
+
+    pub fn stop(&mut self) -> io::Result<()> {
+        self.set_stopped(true);
+        self.with_inner(|inner| inner.stop())
+    }
+}
+
+impl<T> Event<SyncWeak<T>> {
+    fn with_inner<F, O>(&self, f: F) -> O
+    where
+        F: Fn(&mut EventInner<T>) -> O,
+    {
+        let upgraded = self.inner.0.upgrade().unwrap();
+        let mut ev = upgraded.lock().unwrap();
+        f(&mut *ev)
+    }
+
+    pub fn stop(&mut self) -> io::Result<()> {
+        self.set_stopped(true);
+        self.with_inner(|inner| inner.stop())
+    }
+}
+
+impl<T> Drop for EventInner<T> {
+    fn drop(&mut self) {
+        self.drop_context();
```

Comment:
Given that we have `impl<S: Clone> Clone for Event<S>` which actually clones the `inner` I'm wondering - is it safe for `libevent_sys::event_free(raw.as_ptr())` and `finalizer` to be called multiple times for the same event?  Or is something guarding against that occurring?

Reply:
That's a case of me being lazy with `inner` field naming (open to alternative names to differentiate). `Event<S>`'s inner should only ever be the `S` sync wrapper (intentionally precluding `Internal`), so if `EventInner` is getting cloned that's definitely a bug.

Reply:
Ah I see that now - just glossed over this during initial pass.  Thanks for the clarification.